### PR TITLE
Add layouts

### PR DESCRIPTION
**Pull request by griswaldbrooks**

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -17,48 +19,375 @@ struct window {
...
+	  // Push final point
+	  std::vector<int> temp {p1[0], p1[1]};
+
+	  v.push_back(temp);
+
+	  return v;
+
+	}
+
+	std::vector<Cell> bresenham_conversion(const Cell &start, const Cell &end){
```
Comment (briancairl) [inline src/map.cpp]: probably don't need this go between of `std::vector<int>` -> `Cell`, at least not that I can tell. Storing a point as a `std::vector<int>` might get some strange looks, anyway.

File: src/map.cpp [hunk 2]
```diff
@@ -17,48 +19,375 @@ struct window {
...
+
+		const auto pixels = bresenham_conversion(start.x, start.y, end.x, end.y);
+
+		std::vector<Cell> cells;
+
+		for (const auto &pixel:pixels) {
```
Comment (briancairl) [inline src/map.cpp]: but if you do stick with it, maybe this is clearer?
```c++
std::transform(
  std::begin(pixels),
  std::end(pixels),  
  std::back_inserter(cells),
  [](const auto& v) -> Cell
  {
    return {v[0], v[1]};
  }
);
```

File: src/map.cpp [hunk 3]
```diff
@@ -17,48 +19,375 @@ struct window {
...
+        assert(extents_.extent(0) == extents_.extent(1));
+      }
+    }
+
+    constexpr const extents_type& extents() const noexcept {
+      return extents_;
+    }
+
+    constexpr size_type operator()(size_type i) const noexcept {
+      std::array<std::array<std::size_t, 2>, 8> idx {{
```
Comment (briancairl) [inline src/map.cpp]: `static constexpr std::array<std::array<std::size_t, 2>, 8> kIdx` or however you usually flavor your constants.

File: src/map.cpp [hunk 4]
```diff
@@ -17,48 +19,375 @@ struct window {
...
+        {0, 1},
+        {0, 2},
+        {1, 0},
+        {1, 2},
+        {2, 0},
+        {2, 1},
+        {2, 2}
+      }};
+      std::cout << "i = " << i << std::endl;
+      auto coord = idx[i];
```
Comment (briancairl) [inline src/map.cpp]: `const`

File: src/map.cpp [hunk 5]
```diff
@@ -17,48 +19,375 @@ struct window {
...
+}
+
+}  // namespace geometry
+struct bounds_checked_layout_policy {
+  template <typename Extents>
+  struct mapping : std::layout_right::mapping<Extents> {
+    using base_t = std::layout_right::mapping<Extents>;
+    using base_t::base_t;
+    std::ptrdiff_t operator()(auto... idxs) const {
+      [&]<size_t... Is>(std::index_sequence<Is...>) {
```
Comment (briancairl) [inline src/map.cpp:217]: i love that these templated lambdas exist now ❤️

**Review by briancairl — COMMENTED** [review]
some quick stuff.

Is there anything in particular you want feedback-wise?

File: src/map.cpp (hunk 1 above)
Comment (griswaldbrooks) [reply src/map.cpp]: Ah, this is from https://github.com/gillwald/geometry/tree/master
I would refactor it but it's not the focus of my presentation.

File: src/map.cpp (hunk 2 above)
Comment (griswaldbrooks) [reply src/map.cpp]: Sure. I'll do it for now. But like you said, probably don't need two different functions. @bgill92 was just learning how to make his own types.

File: src/map.cpp (hunk 3 above)
Comment (griswaldbrooks) [reply src/map.cpp]: This is pretty old? What commit is this from?

File: src/map.cpp (hunk 5 above)
Comment (griswaldbrooks) [reply src/map.cpp:217]: Yeah. @dhollman did this for me last year.

File: src/map.cpp [hunk 6]
```diff
@@ -17,115 +20,534 @@ struct window {
...
+  int deltas[2] = {(x1 - x0), (y1 - y0)};
+
+  // Absolute values of  the changes
+  int changes[2] = {abs(deltas[0]), abs(deltas[1])};
+
+  // Lambda that emulates signum
+  auto sgn = [](auto a) { return (a > 0) - (a < 0); };
+
+  // Which direction to step towards for the x and y functions
+  int steps[2] = {sgn(deltas[0]), sgn(deltas[1])};
```
Comment (bgill92) [inline src/map.cpp]: Pretty sure all of these can be `const`

File: src/main.cpp [hunk 7]
```diff
@@ -62,7 +57,8 @@ struct bounds_checked_layout_policy {
         if (((idxs < 0 || idxs > this->extents().extent(Is)) || ...)) {
           throw std::out_of_range("Invalid bin index");
         }
-      }(std::make_index_sequence<sizeof...(idxs)>{});
+      }
+      (std::make_index_sequence<sizeof...(idxs)>{});
```
Comment (bgill92) [inline src/main.cpp:60]: You have the `bounds_checked_layout_policy` struct in `map.cpp` also. Maybe put it in a header file?

File: src/map.cpp [hunk 8]
```diff
@@ -17,115 +20,534 @@ struct window {
...
-    // std::advance(first, offset);
-    // return stdex::mdspan(first, window.height, window.width);
-  // }
+  /**
+   * @brief If requested window would go out of bounds, a smaller inbounds map
+   * is returned
+   */
+  subgrid window_from_layout_stride(window size, coordinate corner) {
+    // If the corner is not within bounds, then there is nothing we can do
+    // return std expected error
```
Comment (bgill92) [inline src/map.cpp]: Is this still something you want to do?

File: src/map.cpp [hunk 9]
```diff
@@ -17,115 +20,534 @@ struct window {
...
-    // auto const offset = coord.x * N + coord.y;
-    // auto first = data_.begin();
-    // std::advance(first, offset);
-    // return stdex::mdspan(first, window.height, window.width);
-  // }
+  /**
+   * @brief If requested window would go out of bounds, a smaller inbounds map
+   * is returned
+   */
+  subgrid window_from_layout_stride(window size, coordinate corner) {
```
Comment (bgill92) [inline src/map.cpp]: Maybe `top_left_corner` for name?

Also a small diagram would be helpful

File: src/map.cpp [hunk 10]
```diff
@@ -17,115 +20,534 @@ struct window {
...
+
+    // Shrink window if it goes out of bounds
+    size.width = std::clamp(size.width, std::size_t{0}, Width - corner.x);
+    size.height = std::clamp(size.height, std::size_t{0}, Height - corner.y);
+    // width, height of submap
+    auto const shape = std::dextents<std::size_t, 2>{size.width, size.height};
+    // spacing between elements, spacing between rows
+    // 0 means all of the elements in a row are on top of each other
+    // spacing between rows needs to match the width of the big map
+    auto const strides = std::array<std::size_t, 2>{1, Width};
```
Comment (bgill92) [inline src/map.cpp]: Explain strides better, because wtf.

**Review by bgill92 — COMMENTED** [review]
First pass

File: src/map.cpp (hunk 6 above)
Comment (griswaldbrooks) [reply src/map.cpp]: Quite a thing looking your oldself in the mirror.

File: src/main.cpp (hunk 7 above)
Comment (griswaldbrooks) [reply src/main.cpp:60]: i think I'll end up just having two files for presentation purposes, one for layout and one for accessor examples. I'm not yet turning this into a library

File: src/map.cpp (hunk 8 above)
Comment (griswaldbrooks) [reply src/map.cpp]: no

Comment (griswaldbrooks) [reply src/map.cpp]: I decided clamping it makes more sense

File: src/main.cpp [hunk 11]
```diff
@@ -74,25 +70,28 @@ struct bin_checker {
...
 
   reference access(data_handle_type ptr, std::ptrdiff_t offset) const {
-    // We know ptr will be valid asynchronously because we construct it on the stack
-    // of main. In real code we might want to use a shared_ptr or something
-    return std::async([=]{
-      return bin_state{ptr->is_bin_occupied(static_cast<int>(offset)), offset_to_coord(static_cast<int>(offset))};
+    // We know ptr will be valid asynchronously because we construct it on the
+    // stack of main. In real code we might want to use a shared_ptr or
+    // something
+    return std::async([=] {
```
Comment (briancairl) [inline src/main.cpp:75]: bold and daring of you

Comment (griswaldbrooks) [reply src/main.cpp:75]: Why so?

Comment (briancairl) [reply src/main.cpp:75]: Its implementation-defined, but I think `std::async` usually launches a new thread, so its sort of doodoo. But there's no read/write sync on the data pointed to by ptr. I would guess your `serial::Serial` interfaces do some of their own protection, calling into the same `offset` could be weird?

File: src/map.cpp [hunk 12]
```diff
@@ -491,6 +557,22 @@ requires(Container::extents_type::rank() ==
...
 
+template <std::size_t Height, std::size_t Width>
+void save_image(std::filesystem::path const& filename,
+                occupancy_grid<Height, Width>& grid) {
+  // Create a cv::Mat object of the right size (assuming grayscale image)
+  cv::Mat img(Height, Width, CV_8UC1);  // 8-bit unsigned single-channel
+
+  // Populate the Mat with the data from the array (scale from int to 8-bit
+  // range)
+  for (auto i = 0u; i < Width * Height; ++i) {
```
Comment (briancairl) [inline src/map.cpp]:
```cpp
std::transform(
  std::begin(grid),
  std::end(grid),
  img.begin<uchar>(),
  [](auto grid_value) -> uchar
  {
    return static_cast<uchar>(std::clamp(16 * grid_value, 0, 255));
  } 
);
```
less loopy index-y, but you probably need `grid.begin()` and `grid.end()`

Comment (griswaldbrooks) [reply src/map.cpp]: or i could do neither of those things

**Pull request merged**